src/irr/asset/CImageWriterOpenEXR.cpp

```diff
@@ -0,0 +1,159 @@
+		using namespace IMF;
+		using namespace IMATH;
+
+		constexpr uint8_t availableChannels = 4;
+
+		template<typename ilmType>
+		void createAndWriteImage(std::array<ilmType*, availableChannels>& pixelsArrayIlm, const asset::ICPUImage* image, const char* fileName)
+		{
+			auto getIlmType = [&]()
+			{
+				if (image->getCreationParameters().format == EF_R16G16B16A16_SFLOAT)
+					return PixelType::HALF;
+				else if (image->getCreationParameters().format == EF_R32G32B32A32_SFLOAT)
+					return PixelType::FLOAT;
+				else if (image->getCreationParameters().format == EF_R32G32B32A32_UINT)
+					return PixelType::UINT;
+			};
+
+			Header header(image->getCreationParameters().extent.width, image->getCreationParameters().extent.height);
+			const PixelType pixelType = getIlmType();
+			FrameBuffer frameBuffer;
+
+			const auto& width = image->getCreationParameters().extent.width;
+			const auto& height = image->getCreationParameters().extent.height;
+			const auto pitch = image->getRegions().begin()->bufferRowLength;
+
+			for (uint8_t channel = 0; channel < availableChannels; ++channel)
+				pixelsArrayIlm[channel] = _IRR_NEW_ARRAY(ilmType, width * height);
+
+			for (uint64_t yPos = 0; yPos < height; ++yPos)
+				for (uint64_t xPos = 0; xPos < width; ++xPos)
+				{
+					const uint64_t ptrStyleEndShiftToImageDataPixel = (yPos * pitch * availableChannels) + (xPos * availableChannels);
+					const uint64_t ptrStyleIlmShiftToDataChannelPixel = (yPos * width) + xPos;
+
+					for (uint8_t channelIndex = 0; channelIndex < availableChannels; ++channelIndex)
+					{
+						ilmType channelPixel = *(reinterpret_cast<const ilmType*>(image->getBuffer()->getPointer()) + ptrStyleEndShiftToImageDataPixel + channelIndex);
+						*(pixelsArrayIlm[channelIndex] + ptrStyleIlmShiftToDataChannelPixel) = channelPixel;
+					}
+				}
+
+			constexpr std::array<char*, availableChannels> rgbaSignatureAsText = { "R", "G", "B", "A" };
+			for (uint8_t channel = 0; channel < rgbaSignatureAsText.size(); ++channel)
+			{
+				header.channels().insert(rgbaSignatureAsText[channel], Channel(pixelType));
+				frameBuffer.insert
+				(
+					rgbaSignatureAsText[channel],                                                                // name
+					Slice(pixelType,                                                                             // type
+					(char*) pixelsArrayIlm[channel],                                                             // base
+					sizeof(*pixelsArrayIlm[channel]) * 1,                                                        // xStride
+					sizeof(*pixelsArrayIlm[channel]) * image->getCreationParameters().extent.width)              // yStride
+				);
+			}
+
+			OutputFile file(fileName, header);
+			file.setFrameBuffer(frameBuffer);
+			file.writePixels(image->getCreationParameters().extent.height);
+
+			//for (uint8_t channel = 0; channel < availableChannels; ++channel)
+				// _IRR_DELETE_ARRAY(pixelsArrayIlm[channel], width * height); doesn't work (?)
```

Comment:
what do you mean by doesn't work? doesn't compile?

The macro might not like the pointer type going in (is it a void* ?)

Reply:
it doesn't compile, and type is known by the template

Reply:
and it isn't void*

Reply:
errors?

Reply:
![image](https://user-images.githubusercontent.com/34793522/71409137-c69a8800-2640-11ea-9d0b-d7204743ba6b.png)

Reply:
try making the pointer into a constant pointer and let me know (compiler might get confused).

Reply:
ok I might have fucked this one up in the following lines
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/include/irr/core/memory/new_delete.h#L104

I use `std::remove_pointer` but I should also then wrap it in `std::remove_reference` first.

The compiler is moaning about the fact that the type being passed into the templates is a reference to a type (after you put a pointer type into a `std::remove_pointer` you get a reference type, not the straight vanilla type).

So feel free to correct my macros ;)

Reply:
Macros fixed, but it still doesn't work, but it works with & and works if you create `ilmType* poitner` and try to delete it

Reply:
for each loop works as well

Reply:
curious is that if you perform for each loop when an element is a reference pointer while deallocating memory, it spreads out some errors

Reply:
so its done or not?

Reply:
It's done, above shows you the behaviour